sdks/java/core/src/main/java/org/apache/beam/sdk/testing/PAssert.java

```diff
@@ -121,6 +124,32 @@ private static String nextAssertionName() {
   private PAssert() {}
 
   /**
+   * A {@link DoFn} that counts the number of successful {@link SuccessOrFailure} in the
+   * input {@link PCollection} and counts them. If a failed {@link SuccessOrFailure} is
+   * encountered, it is counted and immediately raised.
+   */
+  private static final class DefaultConcludeFn extends DoFn<SuccessOrFailure, Void> {
+
+    @ProcessElement
+    public void processElement(ProcessContext c) {
+      SuccessOrFailure e = c.element();
+      if (e.isSuccess()) {
+        PAssert.successCounter.inc();
+      } else {
+        PAssert.failureCounter.inc();
+        throw e.assertionError();
+      }
+    }
+  }
+
+  static class DefaultConcludeTransform
```

Comment:
I think this should be public, with Javadoc indicating that runners should replace this with their transform of choice if they want to conclude in a different way.